lib/pulsar/partitioned_producer.ex

```diff
@@ -0,0 +1,167 @@
+
+  ## Returns
+
+  `{:ok, pid}` - The supervisor PID that manages all partition producer groups
+  `{:error, reason}` - Error if the supervisor failed to start
+  """
+  def start_link(name, topic, partitions, opts \\ []) do
+    client = Keyword.get(opts, :client, @default_client)
+    producer_registry = Pulsar.Client.producer_registry(client)
+
+    Supervisor.start_link(
+      __MODULE__,
+      {name, topic, partitions, opts},
+      name: {:via, Registry, {producer_registry, name}}
+    )
+  end
+
+  @doc """
+  Stops a partitioned producer supervisor and all its child producer groups.
+  """
+  def stop(supervisor_pid, reason \\ :normal, timeout \\ :infinity) do
+    Supervisor.stop(supervisor_pid, reason, timeout)
+  end
+
+  @doc """
+  Gets information about all child producer groups managed by this supervisor.
+
+  Returns a list of `{partition_topic, group_pid}` tuples.
+  """
+  def get_partition_groups(supervisor_pid) do
+    supervisor_pid
+    |> Supervisor.which_children()
+    |> Enum.map(fn {partition_topic, group_pid, :supervisor, _modules} ->
+      {partition_topic, group_pid}
+    end)
+  end
+
+  @doc """
+  Gets all producer processes from all partition groups managed by this supervisor.
+
+  Returns a flat list of producer process PIDs from all partitions.
+  """
+  def get_producers(supervisor_pid) do
+    supervisor_pid
+    |> get_partition_groups()
+    |> Enum.flat_map(fn {_partition_topic, group_pid} ->
+      Pulsar.ProducerGroup.get_producers(group_pid)
+    end)
+  end
+
+  @doc """
+  Sends a message through the partitioned producer.
+
+  Routes the message to a partition based on the `:partition_key` option:
+  - With `:partition_key` - Uses consistent hashing to route to a specific partition
+  - Without `:partition_key` - Uses round-robin distribution
+
+  ## Parameters
+
+  - `supervisor_pid` - The partitioned producer supervisor PID
+  - `message` - Binary message payload
+  - `opts` - Optional parameters (same as `Pulsar.ProducerGroup.send_message/3`)
+
+  ## Returns
+
+  `{:ok, message_id}` on success
+  `{:error, reason}` on failure
+  """
+  @spec send_message(pid(), binary(), keyword()) :: {:ok, map()} | {:error, term()}
+  def send_message(supervisor_pid, message, opts \\ []) do
+    partition_groups = get_partition_groups(supervisor_pid)
+    num_partitions = length(partition_groups)
+
+    if num_partitions == 0 do
+      {:error, :no_producers_available}
+    else
+      partition_index = select_partition(supervisor_pid, opts, num_partitions)
+      # Sort by topic name to ensure consistent ordering, then get by index
+      {_topic, group_pid} = partition_groups |> Enum.sort() |> Enum.at(partition_index)
+
+      Pulsar.ProducerGroup.send_message(group_pid, message, opts)
+    end
+  end
+
+  @impl true
+  def init({name, topic, partitions, opts}) do
+    # Create ETS table owned by this supervisor for round-robin counter
+    :ets.new(ets_table_name(self()), [:named_table, :public, :set])
```

Comment:
I am wondering if we could implement this without spinning up a separate ETS table. This would require changes to the partitioned producer, which is currently a supervisor and we would probably need two components, a supervisor but also a process that could orchestrate requests. I would like to think a bit more about this, since it is related to some changes I was considering for consumer groups and partitioned consumers. One option is to start with an ETS table an pivot towards an alternative implementation in the future.